### ISSUE TITLE
Include user provided session in dvc.api

### ISSUE THREAD

**Issue by ElisaSuarez** [iterative/dvc#7710]
Sagemaker uses session objects to interact with other AWS resources. This includes S3 buckets, which in case of Sagemaker's Jupyter Instances use IAM roles to know which buckets it can or cannot access, and it doesn't allow the creation of user profiles to pass to dvc.

Allowing user to directly provide a session (SM uses `boto3.session.Session`, instead of `aiboto3.session.AiSession`) would provide a direct way to use DVC from AWS' Sagemaker.

I am promoting the use of the tool inside my team, so it would be really helpful if this could work in our main work environment which are sagemaker instances.

Thanks a lot!

**Comment by shcheklein** [comment]
@ElisaSuarez could you please show a code example how it's done w/o DVC with these sessions? What kind of information do you provide there? 

Also, how do use `dvc.api` itself.

**Comment by ElisaSuarez** [comment]
There are more than one way to download and upload files to S3 from SageMaker. The more straigthforward one is:

`
from sagemaker.session import Session
from sagemaker.s3 import S3Downloader, S3Uploader

sagemaker_session = Session()
S3Downloader.download(
s3_uri, # The uri of the folder or file to download
local_path, # Local path to put the file or folder
sagemaker_session, # The Session Object
)

S3Uploader.upload(
local_path, # File/folder that will be uploaded
desired_s3, # Prefix of where the file/folder will be added
sagemaker_session,
)
`

The Session object has the information and credentials needed to write/read the S3 to which the notebook has access.

**Comment by shcheklein** [comment]
@ElisaSuarez got it, thanks, but you also mentioned `boto3.session.Session`. Is it the same as as `sagemaker.session.Session`? Do you suggest `dvc.api` to support SM specific session objects?

**Comment by ElisaSuarez** [comment]
The SM session takes an boto3.session.Session object as input, if none is specified it creates one. When you use boto3 library directly, you use a client object.
```
from boto3 import Client

client = boto3.client('s3')

client.download_file(
   Bucket, # Name of the bucket to download from
   Key, # Name of the key to download from
   filename, # Path to the file to download to
)

client.upload_file(
   Filename, # File to upload
   Bucket, # Name of the bucket to upload to
   Key, # Name of the key to upload to
)

```

You can see the documentation for `boto3.client` [here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.upload_file), and for sagemaker session [here](https://sagemaker.readthedocs.io/en/stable/api/utility/session.html#sagemaker.session.Session). I'm paying close attention to this thread so feel free to ask any other question.

Thanks!

**Comment by shcheklein** [comment]
@ElisaSuarez thanks! I still struggle to understand the issue, to be honest. SM session is a bit too specific to pass it into `dvc.api`, but probably it's not even needed - regular boto3 session should be enough (?) since DVC doesn't access any other resources.

But `dvc.api` should be initializing S3 session internally in some way (`aiboto3.session` or something else, I can check). What is the issue do you see with it, why do we need to pass a custom session from a SM machine in the first place?

**Comment by ElisaSuarez** [comment]
As it is right now, DVC looks for the AWS CLI credentials in the `~\.aws\config` file. But SageMaker does not allow for user profiles. When you create a Notebook instance, you assign it an IAM role which grant it access to specific resources (You can give it access only access to one bucket, instead of all the buckets of your account). This results in a credentials error, which we have no way to mitigate.

I think passing a boto3.session.Session, should be enough, also. Or specifically an s3 resource object, if you think is best.

**Comment by pmrowla** [comment]
Can you share the actual error you get when you are trying to use DVC right now?

This issue probably belongs upstream in either s3fs or aiobotocore, on aws systems that have configured IAM credentials, s3fs (and DVC by extension) should already be using those automatically without DVC needing to pass a direct session object into s3fs

**Comment by karajan1001** [comment]
As mentioned above, `Session` object in `sagemaker` is an instance of `boto3.session.Session`.

In DVC we use `s3fs` from `fsspec` as the backend to communicate with the S3 server. It can accept a user's custom session  when initialized 

https://github.com/fsspec/s3fs/blob/8abb586d10c5e0d5ab5020355065d1466138f990/s3fs/core.py#L265

But, what it can only accept an `aioboto3.session.AioSession` instance. And in DVC we also lack a way to pass this session object to the `s3fs`

**Comment by shcheklein** [comment]
> As it is right now, DVC looks for the AWS CLI credentials in the ~\.aws\config file.

That's not exactly true as far as I understand. As @pmrowla and @karajan1001 mentioned it is initializing a session object. And it should be taking into account all the possible ways environment could be initialized (at least the regular one Session) - it probably checks env vars, config files (global, local, etc). Question that we need to understand here - what is the difference between SM initializing it (e.g. SM reads some SM specific env vars and would pass them to create a session) and DVC initializing that session.

**Comment by skshetry** [comment]
This is what Sagemaker does when creating a client:

https://github.com/aws/sagemaker-python-sdk/blob/5b8eb10bfddcb54fb65c8b28f923697619672120/src/sagemaker/session.py#L146

So, it does not look any different than what DVC/s3fs do, maybe some config parsing issue on DVC/s3fs side before creating a session (maybe they are sending unnecessary arguments to it).

**Comment by ElisaSuarez** [comment]
> Can you share the actual error you get when you are trying to use DVC right now?
> 
> This issue probably belongs upstream in either s3fs or aiobotocore, on aws systems that have configured IAM credentials, s3fs (and DVC by extension) should already be using those automatically without DVC needing to pass an alternate session object into s3fs

Instance characteristics that may be relevant:
- Amazon Linux V2
- Using IAM role with permissions to read/write to a single bucket (It only contains the data uploaded by DCV from my local machine)
- This is a Jupyter Notebook instance created from the SageMaker Managment Console
- Both the Jupyter Instance and S3 bucket are in the same region

I start by installing DVC on the instance (this have to be done everytime you shut down the instance)
```
!pip install dvc
```
Which succeeds but generate the following conflicts

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
docker-compose 1.29.2 requires PyYAML<6,>=3.10, but you have pyyaml 6.0 which is incompatible.
awscli 1.22.97 requires botocore==1.24.42, but you have botocore 1.24.19 which is incompatible.
awscli 1.22.97 requires PyYAML<5.5,>=3.10, but you have pyyaml 6.0 which is incompatible.
Successfully installed aiohttp-retry-2.4.6 asyncssh-2.10.1 atpublic-3.0.1 commonmark-0.9.1 configobj-5.0.6 dictdiffer-0.9.0 diskcache-5.4.0 dpath-2.0.6 dulwich-0.20.35 dvc-2.10.2 dvc-render-0.0.5 dvclive-0.7.3 flatten-dict-0.4.2 flufl.lock-7.0 ftfy-6.1.1 funcy-1.17 gitdb-4.0.9 gitpython-3.1.27 grandalf-0.6 mailchecker-4.1.16 nanotime-0.5.2 phonenumbers-8.12.47 psutil-5.9.0 pydot-1.4.2 pygit2-1.9.1 pygtrie-2.4.2 python-benedict-0.25.1 python-fsutil-0.6.0 python-slugify-6.1.2 pyyaml-6.0 rich-12.3.0 ruamel.yaml-0.17.21 ruamel.yaml.clib-0.2.6 scmrepo-0.0.19 shortuuid-1.0.8 shtab-1.5.4 smmap-5.0.0 tqdm-4.64.0 voluptuous-0.13.1 xmltodict-0.12.0 zc.lockfile-2.0
```

With this I create a Repo object and try to push:

```
from dvc.repo import Repo
repo = Repo(PATH)
repo.push()
```

And it generates the following error

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_28280/4045963624.py in <cell line: 1>()
----> 1 repo.push()

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/repo/__init__.py in wrapper(repo, *args, **kwargs)
     46     def wrapper(repo, *args, **kwargs):
     47         with lock_repo(repo):
---> 48             return f(repo, *args, **kwargs)
     49 
     50     return wrapper

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/repo/push.py in push(self, targets, jobs, remote, all_branches, with_deps, all_tags, recursive, all_commits, run_cache, revs, glob, odb, include_imports)
     66                 continue
     67             try:
---> 68                 pushed += self.cloud.push(
     69                     obj_ids, jobs, remote=remote, odb=odb or dest_odb
     70                 )

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/data_cloud.py in push(self, objs, jobs, remote, odb)
     83         if not odb:
     84             odb = self.get_remote_odb(remote, "push")
---> 85         return transfer(
     86             self.repo.odb.local,
     87             odb,

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/data/transfer.py in transfer(src, dest, obj_ids, jobs, verify, hardlink, **kwargs)
    152         return 0
    153 
--> 154     status = compare_status(
    155         src, dest, obj_ids, check_deleted=False, jobs=jobs, **kwargs
    156     )

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/data/status.py in compare_status(src, dest, obj_ids, log_missing, check_deleted, src_index, dest_index, jobs, **kwargs)
    178     if "cache_odb" not in kwargs:
    179         kwargs["cache_odb"] = src
--> 180     dest_exists, dest_missing = status(
    181         dest, obj_ids, index=dest_index, jobs=jobs, **kwargs
    182     )

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/data/status.py in status(odb, obj_ids, name, index, cache_odb, shallow, jobs, **kwargs)
    149         with QueryingProgress(phase="Checking", name=odb.fs_path) as pbar:
    150             exists.update(
--> 151                 odb.hashes_exist(hashes, jobs=jobs, progress=pbar.callback)
    152             )
    153     return StatusResult(

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/objects/db.py in hashes_exist(self, hashes, jobs, progress)
    401 
    402         estimator_cb = partial(progress, "estimating", None)
--> 403         remote_size, remote_hashes = self._estimate_remote_size(
    404             hashes, progress=estimator_cb
    405         )

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/objects/db.py in _estimate_remote_size(self, hashes, progress)
    270             hashes = self._list_hashes(prefix)
    271 
--> 272         remote_hashes = set(iter_with_pbar(hashes))
    273         if remote_hashes:
    274             remote_size = total_prefixes * len(remote_hashes)

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/objects/db.py in iter_with_pbar(hashes)
    258         def iter_with_pbar(hashes):
    259             total = 0
--> 260             for hash_ in hashes:
    261                 total += total_prefixes
    262                 progress(total)

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/objects/db.py in _hashes_with_limit(self, limit, prefix)
    224     def _hashes_with_limit(self, limit, prefix=None):
    225         count = 0
--> 226         for hash_ in self._list_hashes(prefix):
    227             yield hash_
    228             count += 1

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/objects/db.py in _list_hashes(self, prefix)
    214         will be returned.
    215         """
--> 216         for path in self._list_paths(prefix):
    217             try:
    218                 yield self._path_to_hash(path)

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/objects/db.py in _list_paths(self, prefix)
    198         if len(prefix) > 2:
    199             parts = *parts, prefix[2:]
--> 200         yield from self.fs.find(self.fs.path.join(*parts), prefix=bool(prefix))
    201 
    202     def _path_to_hash(self, path):

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/fs/base.py in find(self, path, prefix)
    489         if prefix:
    490             with_prefix = self.path.parent(path)
--> 491             files = self.fs.find(with_prefix, prefix=self.path.parts(path)[-1])
    492         else:
    493             with_prefix = path

~/anaconda3/envs/python3/lib/python3.8/site-packages/funcy/objects.py in __get__(self, instance, type)
     48 
     49                 with ctx:
---> 50                     return prop.__get__(instance, type)
     51 
     52             if hasattr(prop, '__set__'):

~/anaconda3/envs/python3/lib/python3.8/site-packages/funcy/objects.py in __get__(self, instance, type)
     26         if instance is None:
     27             return self
---> 28         res = instance.__dict__[self.fget.__name__] = self.fget(instance)
     29         return res
     30 

~/anaconda3/envs/python3/lib/python3.8/site-packages/dvc/fs/s3.py in fs(self)
    149         from s3fs import S3FileSystem as _S3FileSystem
    150 
--> 151         return _S3FileSystem(**self.fs_args)
    152 
    153     @classmethod

~/anaconda3/envs/python3/lib/python3.8/site-packages/fsspec/spec.py in __call__(cls, *args, **kwargs)
     74             return cls._cache[token]
     75         else:
---> 76             obj = super().__call__(*args, **kwargs)
     77             # Setting _fs_token here causes some static linters to complain.
     78             obj._fs_token_ = token

~/anaconda3/envs/python3/lib/python3.8/site-packages/s3fs/core.py in __init__(self, anon, key, secret, token, use_ssl, client_kwargs, requester_pays, default_block_size, default_fill_cache, default_cache_type, version_aware, config_kwargs, s3_additional_kwargs, session, username, password, **kwargs)
    169         self.s3_additional_kwargs = s3_additional_kwargs or {}
    170         self.use_ssl = use_ssl
--> 171         self.s3 = self.connect()
    172         self._kwargs_helper = ParamKwargsHelper(self.s3)
    173         super().__init__()

~/anaconda3/envs/python3/lib/python3.8/site-packages/s3fs/core.py in connect(self, refresh)
    223                           **self.config_kwargs)
    224             if not self.passed_in_session:
--> 225                 self.session = boto3.Session(self.key, self.secret, self.token,
    226                                              **self.kwargs)
    227 

TypeError: __init__() got an unexpected keyword argument 'profile'
```

and if I try to push from command line:

```
sh-4.2$ cd Sagemaker/my_repo
sh-4.2$ dvc push
ERROR: unexpected error - __init__() got an unexpected keyword argument 'skip_instance_cache'
```

**Comment by ElisaSuarez** [comment]
Thank you so so much! I feel kinda dumb now. I made various tests for pushing from SM, pulling into local, and vice-versa and worked every time.

I only have one last comment on this: I have been mostly following the DVC Getting started tutorials and [in this one](https://dvc.org/doc/start/data-management#storing-and-sharing) says to `git add .dcv/.config`, and this file will be different for each user... When I made the changes and pushed from sagemaker, it also changed the `.config` file, so the profile was overwritten from the name I gave when configuring the aws cli with SSO back to `default`, which in turns force me to always have to run `dvc modify remote myremote --profile myprofile` when pulling in my local machine after a `git pull` made from other machine. Is there a recommended way to deal with this? Specially when more than one person may be working with the data and the source code? (each person modifying the profile's name to reflect their own aws cli setup?
